### PR TITLE
Wrap registered widgets in callbacks to track widget validation errors

### LINKED ISSUE #1005: Add reporting for validation errors coming from widgets
Just as we are now detect the specific shortcode that is the source for an AMP validation error in #971 (and as we'll do for blocks in Gutenberg), we also need to do the same for widgets. When a widget renders something that not valid AMP, we need to indicate:

1. The widget that did something invalid.
2. The theme/plugin source for where the widget came from.

### PR DESCRIPTION
**Pull request by westonruter**
* Include core as a source in addition to plugins and themes.
* Refactor source comments to only consist of JSON-encoded source data.
* Include the function name in the source.

Example validation data for when a Custom HTML widget is added which includes a `script` tag:

image

Fixes #1005.

### CODE REVIEW

**Comment by westonruter** [comment]
- [ ] Add some tests specifically for registered widgets.
- [ ] Try using `all` action instead of wrapping all callbacks up-front.

**Comment by westonruter** [comment]
Ready for review!

**Comment by kienstra** [comment]
**Will Review**

Hi @westonruter,
Sorry for the delay here. I'll review this late today (Sunday) if that's alright.

File: tests/test-class-amp-validation-utils.php [hunk 1]
```diff
@@ -68,16 +68,33 @@ class Test_AMP_Validation_Utils extends \WP_UnitTestCase {
...
 		$this->node   = $dom_document->createElement( self::TAG_NAME );
 		AMP_Validation_Utils::reset_validation_results();
+		$this->original_wp_registered_widgets = $GLOBALS['wp_registered_widgets'];
+	}
+
+	/**
+	 * After a test method runs, reset any state in WordPress the test method might have changed.
+	 */
+	public function tearDown() {
+		$GLOBALS['wp_registered_widgets'] = $this->original_wp_registered_widgets; // WPCS: override ok.
```
Comment (kienstra) [inline tests/test-class-amp-validation-utils.php]: Good idea to reset the original global `$wp_registered_widgets`.

Comment (westonruter) [reply tests/test-class-amp-validation-utils.php]: It should be something that gets reset for us automatically, but it wasn't so this is to ensure we reset the global state.

File: includes/utils/class-amp-validation-utils.php [hunk 2]
```diff
@@ -597,47 +598,34 @@ public static function print_edit_form_validation_status( $post ) {
...
-	 * @return string HTML Comment.
-	 */
-	public static function get_source_comment_end( $type, $name ) {
-		return sprintf( '<!--/amp-source-stack:%s:%s-->', $type, $name );
+	public static function get_source_comment( array $source, $is_start = true ) {
+		unset( $source['reflection'] );
+		return sprintf(
+			'<!--%samp-source-stack %s-->',
+			$is_start ? '' : '/',
+			str_replace( '--', '', wp_json_encode( $source ) )
```
Comment (kienstra) [inline includes/utils/class-amp-validation-utils.php]: It's nice how this adds the json-encoded source data to the comment.

**Pull request merged**